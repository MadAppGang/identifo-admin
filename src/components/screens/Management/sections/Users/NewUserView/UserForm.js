import React, { Component } from 'react';
import update from '@madappgang/update-by-path';
import PropTypes from 'prop-types';
import Input from '~/components/shared/Input';
import Field from '~/components/shared/Field';
import Button from '~/components/shared/Button';
import SaveIcon from '~/components/icons/SaveIcon';
import LoadingIcon from '~/components/icons/LoadingIcon';

import './UserForm.css';

class UserForm extends Component {
  constructor() {
    super();

    this.state = {
      fields: {
        name: '',
        email: '',
        password: '',
        confirmPassword: '',
      },
    };

    this.handleInput = this.handleInput.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  handleInput({ target }) {
    this.setState(state => ({
      fields: update(state.fields, {
        [target.name]: target.value,
      }),
    }));
  }

  handleSubmit(event) {
    event.preventDefault();

    const { fields } = this.state;

    this.props.onSubmit(update(fields, 'confirmPassword', undefined));
  }

  render() {
    const { email, name, password, confirmPassword } = this.state.fields;
    const { saving } = this.props;

    return (
      <form className="iap-users-form" onSubmit={this.handleSubmit}>
        <Field label="Name">
          <Input
            name="name"
            value={name}
            placeholder="Enter name"
            onChange={this.handleInput}
          />
        </Field>

        <Field label="Email">
          <Input
            name="email"
            value={email}
            placeholder="Enter email"
            onChange={this.handleInput}
          />
        </Field>

        <Field label="Password">
          <Input
            name="password"
            type="password"
            value={password}
            placeholder="Enter password"
            onChange={this.handleInput}
          />
        </Field>

        <Field label="Confirm Password">
          <Input
            name="confirmPassword"
            type="password"
            value={confirmPassword}
            placeholder="Enter password once more"
            onChange={this.handleInput}
          />
        </Field>

        <footer className="iap-users-form__footer">
          <Button Icon={saving ? LoadingIcon : SaveIcon} type="submit">
            Save User
          </Button>
          <Button transparent onClick={this.props.onCancel}>
            Cancel
          </Button>
        </footer>
      </form>
    );
  }
}

UserForm.propTypes = {
  onCancel: PropTypes.func,
  onSubmit: PropTypes.func,
  saving: PropTypes.bool,
};

UserForm.defaultProps = {
  onCancel: () => {},
  onSubmit: () => {},
  saving: false,
};

export default UserForm;
