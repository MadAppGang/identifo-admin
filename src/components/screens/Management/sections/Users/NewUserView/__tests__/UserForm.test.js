import UserForm from '../UserForm';

describe('new user screen <UserForm />', () => {
  let props;

  beforeEach(() => {
    props = {
      onCancel: jest.fn(),
      onSubmit: jest.fn(),
      saving: false,
      error: null,
    };
  });

  test('renders as expected with no error', () => {
    expect(shallow(<UserForm {...props} />)).toMatchSnapshot();
  });

  test('renders as expected with passed error', () => {
    props.error = new Error('error');
    expect(shallow(<UserForm {...props} />)).toMatchSnapshot();
  });

  test('renders as expected in saving state', () => {
    props.saving = true;
    expect(shallow(<UserForm {...props} />)).toMatchSnapshot();
  });

  test('not empty validation message on "email" field', () => {
    const component = shallow(<UserForm {...props} />);

    const event = {
      target: {
        name: 'email',
        value: '',
      },
    };

    component.find('Input[name="email"]').simulate('blur', event);
    expect(component.find('Input[name="email"]').prop('errorMessage'))
      .toBe('Email should not be empty');
  });

  test('format validation message on "email" field', () => {
    const component = shallow(<UserForm {...props} />);

    const event = {
      target: {
        name: 'email',
        value: 'invalid@email',
      },
    };

    component.find('Input[name="email"]').simulate('blur', event);
    expect(component.find('Input[name="email"]').prop('errorMessage'))
      .toBe('Email format is invalid');
  });

  test('not empty validation message on "password" field', () => {
    const component = shallow(<UserForm {...props} />);

    const event = {
      target: {
        name: 'password',
        value: '',
      },
    };

    component.find('Input[name="password"]').simulate('blur', event);
    expect(component.find('Input[name="password"]').prop('errorMessage'))
      .toBe('Password should not be empty');
  });

  test('not empty validation message on "confirmPassword" field', () => {
    const component = shallow(<UserForm {...props} />);

    const event = {
      target: {
        name: 'confirmPassword',
        value: '',
      },
    };

    component.find('Input[name="confirmPassword"]').simulate('blur', event);
    expect(component.find('Input[name="confirmPassword"]').prop('errorMessage'))
      .toBe('You should confirm password');
  });

  test('passwords dont match validation message on "confirmPassword" field', () => {
    const component = shallow(<UserForm {...props} />);

    const event = {
      target: {
        name: 'confirmPassword',
        value: 'happy',
      },
    };

    component.find('Input[name="confirmPassword"]').simulate('blur', event);
    expect(component.find('Input[name="confirmPassword"]').prop('errorMessage'))
      .toBe('Passwords do not match');
  });

  test('calls onCancel prop function on "cancel" button click', () => {
    shallow(<UserForm {...props} />)
      .find('Button').at(1)
      .simulate('click');

    expect(props.onCancel).toBeCalled();
  });

  test('calls onSubmit prop function on "submit" button click', () => {
    const component = shallow(<UserForm {...props} />);

    const inputData = {
      name: 'Name',
      email: 'john@doe.com',
      password: 'happy123',
      confirmPassword: 'happy123',
    };

    component.find('Input[name="name"]')
      .simulate('change', { target: { name: 'name', value: inputData.name } });

    component.find('Input[name="email"]')
      .simulate('change', { target: { name: 'email', value: inputData.email } });

    component.find('Input[name="password"]')
      .simulate('change', { target: { name: 'password', value: inputData.password } });

    component.find('Input[name="confirmPassword"]')
      .simulate('change', { target: { name: 'confirmPassword', value: inputData.confirmPassword } });

    component.find('form').simulate('submit', { preventDefault: jest.fn() });

    expect(props.onSubmit).toBeCalledWith({ ...inputData, confirmPassword: undefined });
  });
});
