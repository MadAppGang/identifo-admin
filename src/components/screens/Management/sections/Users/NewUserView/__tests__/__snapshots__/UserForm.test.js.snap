// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`new user screen <UserForm /> renders as expected in saving state 1`] = `
<form
  className="iap-users-form"
  onSubmit={[Function]}
>
  <Field
    label="Name"
  >
    <Input
      Icon={null}
      errorMessage=""
      name="name"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Enter name"
      type="text"
      value=""
    />
  </Field>
  <Field
    label="Email"
  >
    <Input
      Icon={null}
      errorMessage=""
      name="email"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Enter email"
      type="text"
      value=""
    />
  </Field>
  <Field
    label="Password"
  >
    <Input
      Icon={null}
      errorMessage=""
      name="password"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Enter password"
      type="password"
      value=""
    />
  </Field>
  <Field
    label="Confirm Password"
  >
    <Input
      Icon={null}
      errorMessage=""
      name="confirmPassword"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Enter password once more"
      type="password"
      value=""
    />
  </Field>
  <footer
    className="iap-users-form__footer"
  >
    <Button
      Icon={[Function]}
      disabled={true}
      error={false}
      iconClassName=""
      onClick={null}
      stretch={false}
      transparent={false}
      type="submit"
    >
      Save User
    </Button>
    <Button
      Icon={null}
      disabled={true}
      error={false}
      iconClassName=""
      onClick={[MockFunction]}
      stretch={false}
      transparent={true}
      type="button"
    >
      Cancel
    </Button>
  </footer>
</form>
`;

exports[`new user screen <UserForm /> renders as expected with no error 1`] = `
<form
  className="iap-users-form"
  onSubmit={[Function]}
>
  <Field
    label="Name"
  >
    <Input
      Icon={null}
      errorMessage=""
      name="name"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Enter name"
      type="text"
      value=""
    />
  </Field>
  <Field
    label="Email"
  >
    <Input
      Icon={null}
      errorMessage=""
      name="email"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Enter email"
      type="text"
      value=""
    />
  </Field>
  <Field
    label="Password"
  >
    <Input
      Icon={null}
      errorMessage=""
      name="password"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Enter password"
      type="password"
      value=""
    />
  </Field>
  <Field
    label="Confirm Password"
  >
    <Input
      Icon={null}
      errorMessage=""
      name="confirmPassword"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Enter password once more"
      type="password"
      value=""
    />
  </Field>
  <footer
    className="iap-users-form__footer"
  >
    <Button
      Icon={[Function]}
      disabled={false}
      error={false}
      iconClassName=""
      onClick={null}
      stretch={false}
      transparent={false}
      type="submit"
    >
      Save User
    </Button>
    <Button
      Icon={null}
      disabled={false}
      error={false}
      iconClassName=""
      onClick={[MockFunction]}
      stretch={false}
      transparent={true}
      type="button"
    >
      Cancel
    </Button>
  </footer>
</form>
`;

exports[`new user screen <UserForm /> renders as expected with passed error 1`] = `
<form
  className="iap-users-form"
  onSubmit={[Function]}
>
  <FormErrorMessage
    error={[Error: error]}
  />
  <Field
    label="Name"
  >
    <Input
      Icon={null}
      errorMessage=""
      name="name"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Enter name"
      type="text"
      value=""
    />
  </Field>
  <Field
    label="Email"
  >
    <Input
      Icon={null}
      errorMessage=""
      name="email"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Enter email"
      type="text"
      value=""
    />
  </Field>
  <Field
    label="Password"
  >
    <Input
      Icon={null}
      errorMessage=""
      name="password"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Enter password"
      type="password"
      value=""
    />
  </Field>
  <Field
    label="Confirm Password"
  >
    <Input
      Icon={null}
      errorMessage=""
      name="confirmPassword"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Enter password once more"
      type="password"
      value=""
    />
  </Field>
  <footer
    className="iap-users-form__footer"
  >
    <Button
      Icon={[Function]}
      disabled={false}
      error={true}
      iconClassName=""
      onClick={null}
      stretch={false}
      transparent={false}
      type="submit"
    >
      Save User
    </Button>
    <Button
      Icon={null}
      disabled={false}
      error={false}
      iconClassName=""
      onClick={[MockFunction]}
      stretch={false}
      transparent={true}
      type="button"
    >
      Cancel
    </Button>
  </footer>
</form>
`;
